roachtest: import/tpch/nodes=32 failed

SHA: https://github.com/cockroachdb/cockroach/commits/395d842feb97c5bd8cad2b32b71a5156c03061eb

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=import/tpch/nodes=32 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1115923&tab=buildLog

```
The test failed on master:
	test.go:743,cluster.go:1585,import.go:150: unexpected node event: 16: dead

```

The input RaftStatus was nil. A comment I wrote claimed that this can't happen:

https://github.com/cockroachdb/cockroach/blob/b8eab7a30993b123ca7c7b9dc2a8b9ca5d7b7bf1/pkg/storage/raft_log_queue.go#L206

It's definitely a lie because we sometimes return trivial `truncatedStatus`es:

https://github.com/cockroachdb/cockroach/blob/b8eab7a30993b123ca7c7b9dc2a8b9ca5d7b7bf1/pkg/storage/raft_log_queue.go#L129-L134

I think `String()` wasn't called on truncateDecisions that don't cause a truncation, but we recently changed that (#34266). Easy to fix.

```
E190130 09:36:20.286516 52416 util/log/crash_reporting.go:202  [n22,raftlog,s22,r786/1:/Table/53/1/1081{0310…-1991…}] a panic has occurred!
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x190242d]

goroutine 52416 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0001c4a20, 0x3898dc0, 0xc005d8be30)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:183 +0x11f
panic(0x2d24dc0, 0x546ff60)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage.(*truncateDecision).raftSnapshotsForIndex(0xc0064887d0, 0x0, 0xc02d188000)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_log_queue.go:226 +0x3d
github.com/cockroachdb/cockroach/pkg/storage.(*truncateDecision).NumNewRaftSnapshots(0xc0064887d0, 0xc006d30e00)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_log_queue.go:254 +0x34
github.com/cockroachdb/cockroach/pkg/storage.(*truncateDecision).String(0xc0064887d0, 0xc008ad5a00, 0xc008184000)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_log_queue.go:273 +0x24a
github.com/cockroachdb/cockroach/pkg/storage.(*raftLogQueue).process(0xc0006362c0, 0x3898d80, 0xc008ad5a40, 0xc008184000, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_log_queue.go:449 +0x3d0
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).processReplica(0xc0009d8000, 0x3898d80, 0xc008ad5a40, 0xc008184000, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:751 +0x345
github.com/cockroachdb/cockroach/pkg/storage.(*baseQueue).processLoop.func1.2(0x3898dc0, 0xc005d8be30)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/queue.go:645 +0xb8
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0001c4a20, 0x3898dc0, 0xc005d8be30, 0xc0070fe0c0, 0x2b, 0x0, 0x0, 0xc006d30de0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:323 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:318 +0x134
~
```

--- a/pkg/storage/raft_log_queue.go
+++ b/pkg/storage/raft_log_queue.go
@@ -90,9 +90,9 @@ func newRaftLogQueue(store *Store, db *client.DB, gossip *gossip.Gossip) *raftLo
 }
 
 // newTruncateDecision returns a truncateDecision for the given Replica if no
-// error occurs. If no truncation can be carried out, a zero decision is
-// returned.
-func newTruncateDecision(ctx context.Context, r *Replica) (*truncateDecision, error) {
+// error occurs. If input data to establish a truncateDecision is missing, a
+// zero decision is returned.
+func newTruncateDecision(ctx context.Context, r *Replica) (truncateDecision, error) {
 	rangeID := r.RangeID
 	now := timeutil.Now()
 
@@ -123,20 +123,20 @@ func newTruncateDecision(ctx context.Context, r *Replica) (*truncateDecision, er
 	r.mu.Unlock()
 
 	if err != nil {
-		return nil, errors.Errorf("error retrieving first index for r%d: %s", rangeID, err)
+		return truncateDecision{}, errors.Errorf("error retrieving first index for r%d: %s", rangeID, err)
 	}
 
 	if raftStatus == nil {
 		if log.V(6) {
 			log.Infof(ctx, "the raft group doesn't exist for r%d", rangeID)
 		}
-		return &truncateDecision{}, nil
+		return truncateDecision{}, nil
 	}
 
 	// Is this the raft leader? We only perform log truncation on the raft leader
 	// which has the up to date info on followers.
 	if raftStatus.RaftState != raft.StateLeader {
-		return &truncateDecision{}, nil
+		return truncateDecision{}, nil
 	}
 
 	// For all our followers, overwrite the RecentActive field (which is always
@@ -155,7 +155,7 @@ func newTruncateDecision(ctx context.Context, r *Replica) (*truncateDecision, er
 	}
 
 	input := truncateDecisionInput{
-		RaftStatus:                     raftStatus,
+		RaftStatus:                     *raftStatus,
 		LogSize:                        raftLogSize,
 		MaxLogSize:                     targetSize,
 		FirstIndex:                     firstIndex,
@@ -164,7 +164,7 @@ func newTruncateDecision(ctx context.Context, r *Replica) (*truncateDecision, er
 	}
 
 	decision := computeTruncateDecision(input)
-	return &decision, nil
+	return decision, nil
 }
 
 func updateRaftProgressFromActivity(
@@ -203,7 +203,7 @@ const (
 )
 
 type truncateDecisionInput struct {
-	RaftStatus                     *raft.Status // never nil
+	RaftStatus                     raft.Status
 	LogSize, MaxLogSize            int64
 	FirstIndex, LastIndex          uint64
 	PendingPreemptiveSnapshotIndex uint64
@@ -315,7 +315,7 @@ func (td *truncateDecision) ShouldTruncate() bool {
 // snapshots. See #8629.
 func computeTruncateDecision(input truncateDecisionInput) truncateDecision {
 	decision := truncateDecision{Input: input}
-	decision.QuorumIndex = getQuorumIndex(input.RaftStatus)
+	decision.QuorumIndex = getQuorumIndex(&input.RaftStatus)
 
 	decision.NewFirstIndex = decision.QuorumIndex
 	decision.ChosenVia = truncatableIndexChosenViaQuorumIndex

--- a/pkg/storage/raft_log_queue_test.go
+++ b/pkg/storage/raft_log_queue_test.go
@@ -201,7 +201,7 @@ func TestComputeTruncateDecision(t *testing.T) {
 			"should truncate: false [truncate 0 entries to first index 2 (chosen via: first index)]",
 		}}
 	for i, c := range testCases {
-		status := &raft.Status{
+		status := raft.Status{
 			Progress: make(map[uint64]raft.Progress),
 		}
 		for j, v := range c.progress {
@@ -243,7 +243,7 @@ func TestComputeTruncateDecisionProgressStatusProbe(t *testing.T) {
 
 	testutils.RunTrueAndFalse(t, "tooLarge", func(t *testing.T, tooLarge bool) {
 		testutils.RunTrueAndFalse(t, "active", func(t *testing.T, active bool) {
-			status := &raft.Status{
+			status := raft.Status{
 				Progress: make(map[uint64]raft.Progress),
 			}
 			for j, v := range []uint64{500, 400, 300, 200, 100} {
@@ -277,10 +277,20 @@ func TestComputeTruncateDecisionProgressStatusProbe(t *testing.T) {
 	})
 }
 
+func TestTruncateDecisionZeroValue(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	var decision truncateDecision
+	assert.False(t, decision.ShouldTruncate())
+	assert.Zero(t, decision.NumNewRaftSnapshots())
+	assert.Zero(t, decision.NumTruncatableIndexes())
+	assert.Equal(t, "should truncate: false [truncate 0 entries to first index 0 (chosen via: )]", decision.String())
+}
+
 func TestTruncateDecisionNumSnapshots(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
-	status := &raft.Status{
+	status := raft.Status{
 		Progress: map[uint64]raft.Progress{
 			// Fully caught up.
 			5: {State: raft.ProgressStateReplicate, Match: 11, Next: 12},

